Create section in buddy list for users from narrow.

evykassirer opened the pull request:
Fixes #21285.

[CZO thread](https://chat.zulip.org/#narrow/stream/101-design/topic/buddy.20list.20split.20users.20.2326717/near/1687871)

Remaining work:

* [x] tooltips on both sections to pointing left, like the tooltips on users's names, since that way they won't block the content of their section
* [x] Add a delay on the tooltips
* [ ] (if easy) change "In this stream" to "In #stream", ideally with the correct privacy icon, as in Vlad's screenshot [here](https://chat.zulip.org/#narrow/stream/101-design/topic/buddy.20list.20split.20users.20.2326717/near/1684091)? If this gets tricky with live updates, or long stream names, etc., let's leave it for a follow-up.
* [x] Inactive users are hidden in PMs again for some reason, so I need to investigate that.
* [x] [starting this conversation in #frontend](https://github.com/zulip/zulip/pull/26717#discussion_r1399937533)
* [x] change triangles to point right when collapsed
* [x] Unit tests

Followups:

* [x] play with rounding function (e.g., Alya thinks she'd prefer 1.5K to 1K for 1509 users), "but that should definitely be a separate follow-up."
* [x] clicking "view all subscribers" opens the personal tab and then the subscribers tab (the URL changes twice) -- #28235 

Waiting on discussion:

* [ ] Finalized triangle design
* [ ] Could we add the name of the stream to the tooltip as well, for clarity? I think "17 subscribers in # stream" (with the correct privacy icon) should work. (Note that we just did something similar in https://github.com/zulip/zulip/pull/27666 .)

<details>
<summary>(oldish) screenshots:</summary>

DMs

<img width="1074" alt="image" src="https://github.com/zulip/zulip/assets/5634097/da8111a3-02ee-458d-8d80-fcd918abc1c8">

stream with some users subscribed:

<img width="1066" alt="image" src="https://github.com/zulip/zulip/assets/5634097/cf6e68bc-d2a1-4f5c-b7c9-6daa0cd00de2">

stream with all users subscribed (no separation)

<img width="1074" alt="image" src="https://github.com/zulip/zulip/assets/5634097/28e766e3-2632-4003-a1c2-a02de7905ee9">

</details>

I've tested it on an organization with 1000 users and it looked okay to me.

To test large organizations, you might need to add this to stop the contact list to shrinking to not have anyone in it:

```
function maybe_shrink_list(user_ids, user_filter_text) {
    return user_ids;
```

evykassirer commented on web/src/buddy_data.js:
```diff
@@ -66,7 +55,27 @@ export function level(user_id) {
 }
 
-export function compare_function(a, b) {
+export function would_receive_message(user_id, current_sub, pm_ids_string) {
```
from Tim earlier:

> Well you probably want a different name then -- something like `user_matches_narrow` or something -- since the name `would_receive_message` was chosen to refer to specifically if that user would receive the message currently being composed.
> 
> I also note that this has an `ids.includes` check in it -- some care may be needed here to use sets and thus avoid accidentally quadratic code.

timabbott commented on web/src/buddy_data.js:
```diff
@@ -66,7 +68,26 @@ export function level(user_id) {
+export function user_matches_narrow(user_id, current_sub, pm_ids) {
+    if (current_sub) {
+        return stream_data.is_user_subscribed(current_sub.stream_id, user_id);
+    }
```
Would a `stream_id` be a better parameter for this function than `current_sub`?

evykassirer replied:
Sure! Changed that.

timabbott commented on web/src/buddy_data.js:
```diff
@@ -86,7 +107,9 @@ export function compare_function(a, b) {
     // TODO sort by unread count first, once we support that
-    user_ids.sort(compare_function);
+    const current_sub = narrow_state.stream_sub();
+    const pm_ids = new Set(narrow_state.pm_ids_string()?.split(","));
```
I think `people.user_ids_string_to_ids_array` might be better here -- we want to try to avoid duplicated parsing code to the extent possible because it's hard to check if it's correct.

evykassirer replied:
Thanks for pointing me to that! The code is a bit longer now but hopefully easier to read. Let me know if you think there's a nice way to clean it up further.

before: `const pm_ids = new Set(narrow_state.pm_ids_string()?.split(","));`

after: 

```js
    const pm_ids_string = narrow_state.pm_ids_string();
    if (pm_ids_string) {
        const pm_ids = new Set(people.user_ids_string_to_ids_array(pm_ids_string));
        user_ids.sort((a, b) => compare_function(a, b, current_sub, pm_ids));
    }
    return user_ids;
```

evykassirer replied:
Thanks for pointing me to that! The code is a bit longer now but hopefully easier to read? I'm not convinced though. Let me know if you think there's a nice way to clean it up further. Maybe `people.user_ids_string_to_ids_array` takes an empty string?

before: `const pm_ids = new Set(narrow_state.pm_ids_string()?.split(","));`

after: 
```js
const pm_ids_string = narrow_state.pm_ids_string();
const pm_ids_list = pm_ids_string ? people.user_ids_string_to_ids_array(pm_ids_string) : [];
const pm_ids_set = new Set(pm_ids_list);
user_ids.sort((a, b) => compare_function(a, b, current_sub, pm_ids_set));
```

timabbott replied:
The resulting code seems OK to me. The other reasonable option would be a more explicit if/else.

timabbott commented on web/src/buddy_list.js:
```diff
@@ -83,12 +88,45 @@ export class BuddyList extends BuddyListConf {
+        $("#narrow-users-presence-container .buddy-list-subsection-header").toggleClass(
+            "no-display",
+            other_users.length === 0,
+        );
```
The subsection headers logic is a bit puzzling to me -- we toggle the other subsection based on it whether this one has rows under it?

evykassirer replied:
I made it so that the headers go away if only one section is visible. 

We definitely should do this for Other Users (since having only one section and calling it "other" is odd). 

I'm open to keeping the narrow users header when other users isn't visible, if folks prefer that.

timabbott commented on web/src/narrow.js:
```diff
@@ -578,6 +579,9 @@ export function activate(raw_operators, opts) {
     }
+    if (activity.user_filter) {
+        activity.build_user_sidebar();
+    }
```
This should be done in `pm_list.handle_narrow_activated`, I think. Similarly for the `deactivate` code path.

evykassirer replied:
Ah yes that makes sense, thanks!

timabbott commented on web/src/recent_view_ui.js:
```diff
@@ -932,6 +933,9 @@ export function show() {
     resize.update_recent_view_filters_height();
+    if (activity.user_filter) {
+        activity.build_user_sidebar();
+    }
```
Similarly I think this goes in `pm_list.handle_narrow_deactivated`.

evykassirer replied:
Ah great, removed this from here

timabbott commented on web/src/buddy_list.js:
```diff
@@ -27,8 +29,10 @@ class BuddyListConf {
+        return $narrow_users_container.find(
+            `${this.item_sel}[data-user-id='${CSS.escape(user_id)}']`,
+        );
     }
```
I don't remember what this function does but it seems like something that should find things in `$other_users_container` too?

timabbott commented on web/src/buddy_list.js:
```diff
@@ -39,9 +47,9 @@ class BuddyListConf {
+        const user_ids = opts.user_ids;
+        const data = buddy_data.get_items_for_users(user_ids);
         return data;
     }
```
Should this function just accept the `user_ids`?

timabbott commented on web/src/buddy_list.js:
```diff
@@ -118,63 +131,89 @@ export class BuddyList extends BuddyListConf {
+        if (i === 0 && this.narrow_user_ids.length > 0) {
+            return this.narrow_user_ids.at(-1);
+        }
+        return undefined;
```
This may have a bug, or maybe just these need better comments -- at least it looks like this logic has going to the "previous" element from the top of `narrow_user_ids` taking you to the bottom of the `narrow_user_ids` block as a wrap-around, and not to the bottom of `other_user_ids`? I'm not sure the intend but it seems inconsistent.

timabbott replied:
@alya probably worth discussing if we want wrap-around to work here, since it's a decent amount of keyboard navigation code to re-review/test either way, and I'd like us to just do whatever version we want to end up with rather than doing it twice.

alya replied:
Hmm, we don't wrap around in the left sidebar, so maybe not wrapping here is most consistent? But probably worth asking in #design, in case folks who use keyboard nav more have an opinion. (And we could change it in the left sidebar, of course.)

evykassirer replied:
I'd prefer to keep it as is for now, and switch to wrapping (in both places) if we want to as a separate task

timabbott replied:
Yeah that seems mostly fine.

timabbott commented on web/src/buddy_list.js:
```diff
@@ -118,63 +131,89 @@ export class BuddyList extends BuddyListConf {
+        let i = this.other_user_ids.indexOf(key);
+        if (i >= 0) {
+            return this.other_user_ids[i + 1];
         }
```
This returns `undefined` when at the bottom -- is that intended?

timabbott commented on web/src/compose_fade.js:
```diff
@@ -1,181 +0,0 @@
-    }
-
-    compose_fade_helper.set_focused_recipient(focused_recipient);
-}
```
You're removing too much here -- `compose_fade` manages both the message feed fading and parts of the buddy list fading (`compose_fade_users`) -- we are only removing the latter feature.

timabbott commented:
Thanks for all the work on this @evykassirer!

I didn't look at the CSS at all and the `prev/next` rewrite logic needs a more careful pass once you've considered the comments above; it's the kind of thing that can easily have bugs. But the structure seems generally reasonable and the above comments should help you clean this up considerably.

Planning to square the temp commits including the test changes is very reasonable.

evykassirer commented:
Did some of a pass, now off to dinner and will finish later :)

evykassirer reviewed (commented):
Addressed everything except re-adding some of the compose fade logic. Thanks for the helpful comments! 

> Planning to square the temp commits including the test changes is very reasonable.

@timabbott can you explain what you mean by square? Is that a typo for squash? (And if so, is it helpful for me to keep the commits unsquashed for now, for reviewing?)

evykassirer commented:
@timabbott I've now addressed all your comments.

One thing I haven't done yet is add more unit tests to bring test coverage back up to 100%.

--

@alya Can you take a look and see how it feels? 

One thing to note is I made it so that the section headers go away if only one section is visible. I think we should definitely do this if all users are in the Other Users section (since having only one section and calling it "other" is odd).

I'm open to keeping the Narrow Users header when there are no other users, if folks prefer that.

If we decide to have just one header sometimes, I'd like some input on what header to show for narrows like Recent Conversations and Starred Conversations, where there aren't really recipients for the narrow. Since they're technically being put in the Other Users bucket, we could decide to just never show Other Users as a single header, but be okay showing Narrow Users as a single header.

Let me know what you think, or if you have clarifying questions -- it's a bit hard/confusing for me to describe this ^^'

evykassirer commented:
Most recent push involves adding new tests to bring test coverage to 100%. I want to do some tidying of them and set up some shared code.

timabbott commented:
> Is that a typo for squash

Yes. Not sure about whether it's useful to keep them unsquashed; I'll look at that in my next review. I think the next step for this is for @alya to do a full product review of it.

alya commented:
Nice, this is feeling pretty good! A few thoughts on UX details and labels:

1. I think we should use a different label than "In selected conversation". One concern I have is that with @-topic mentions and the options for automatically following topics, the notion of being a conversation participant is becoming more prominent, and "In selected conversation" sounds like it might refer to that. (So people who actually participated in the topic, vs. all the stream subscribers.") We could try something like "In this stream"  for streams / "In this conversation" for DMs? 

The issue said "Subscribers" for both contexts, which I'm not sure how I feel about at this point, but @timabbott still likes it. So that's also an option.

2. While playing with it, it felt confusing that the "In this conversation" section label goes away in (a) streams where everyone in the organization is subscribed, and (b) if you filter names. It makes it seem like maybe the feature to show who's in the conversation doesn't exist in some contexts.

I think the simplest and most consistent solution is to just always show both section labels, even if one of the sections is empty. Not showing any labels seems fine when there are no users matching the current filter at all.

3. @timabbott suggested putting the collapse/uncollapse icon on the right rather than the left, to make it possible to left-align the names (which is also how it's shown in the issue). Can we try that?

evykassirer commented:
I wrote some more about #2 [above](https://github.com/zulip/zulip/pull/26717#issuecomment-1738433753)

alya commented:
Thanks! I'm still seeing the headings disappear when filtering, whereas I think we want to keep them around:

<img width="278" alt="Screenshot 2023-10-02 at 1 22 21 PM" src="https://github.com/zulip/zulip/assets/2090066/ad5434ce-bcab-4c1e-8468-d365774ce625">

alya commented:
Oh, we should show the whole list of users in the "All DMs" view:

<img width="1028" alt="Screenshot 2023-10-02 at 1 23 47 PM" src="https://github.com/zulip/zulip/assets/2090066/65a7c92a-1e7e-4301-880f-673f7ca91eda">

alya commented:
This is what I see in a stream where the current user (aaron) is the only subscriber. I would expect to have the two sections as usual:

<img width="282" alt="Screenshot 2023-10-02 at 1 26 04 PM" src="https://github.com/zulip/zulip/assets/2090066/8411ee26-d954-4093-8ecd-cef34aca36d7">

evykassirer commented:
Oh sorry I had some old search-specific code that was messing some stuff up with the recent change. I think removing it fixed all three of those issues -- how's it look now?

alya commented:
I'm still seeing a lack of headers when only a user under "Others" matches the current filter.

alya commented:
"No matching users." feels odd to me when the filter is not being used. Maybe something like "_None_" would work better?

evykassirer commented:
> I'm still seeing a lack of headers when only a user under "Others" matches the current filter.

Ohhhh I see what's going on. Okay hopefully this fixes it. Sorry for the back and forth!

evykassirer commented:
I think I prefer "None" to not having anything there at all. Do you have a preference either way?

alya commented:
> I think I prefer "None" to not having anything there at all. I like the consistency of both headers visually having the arrow and something beneath them.

Happy to try that; seems easy enough to iterate on it when we test on CZO and get feedback from more folks.

alya commented:
The headings seem to stick around like we want them to now!

alya commented:
One design quirk is that "No matching users" has much greater contrast in dark theme than light theme. To me, it feels too contrasty in dark theme, and maybe not contrasty enough in light theme. But maybe we don't need to worry about it given that a right sidebar redesign is coming soon.

![Screenshot 2023-10-03 at 1 42 15 PM](https://github.com/zulip/zulip/assets/2090066/e9af5390-f869-4910-8f59-3578d015c168)
![Screenshot 2023-10-03 at 1 42 05 PM](https://github.com/zulip/zulip/assets/2090066/340975da-e95c-4814-9f95-125ad89a1d8a)

alya commented:
After the "None" change, I guess @timabbott should take another look to decide on next steps.

evykassirer commented:
That push fixes a bug found in a puppeteer test, still need to update the empty list placeholder.

evykassirer commented:
Updated to use the None placeholder outside of search :) Ready for @timabbott now!

evykassirer commented on web/src/buddy_data.js:
```diff
@@ -1,3 +1,5 @@
+import $ from "jquery";
```
@timabbott how do you feel about using jquery in this file? I wasn't sure how else to set the placeholder text only when search changed, while also not causing circular imports

timabbott replied:
It's definitely an abstraction boundary violation. But we can plan to clean that up in a follow-up. Please add a "TODO" comment next to `set_is_searching_users` indicating that it should be in X module but can't be easily placed there due to import cycles.

evykassirer replied:
Done!

karlstolley commented on web/styles/dark_theme.css:
```diff
@@ -784,8 +784,8 @@
 
-    #user_presences li:hover,
-    #user_presences li.highlighted_user {
+    #narrow-user-presences li:hover,
```
This looks like a good opportunity to set CSS variables on the colors here and below, and get further out of the business of duplicate selectors in this dark theme file.

evykassirer replied:
I don't want to do it in the same commit, since that was just renaming selectors, but I'm happy to put a commit on the end of this PR with a variable! Great idea :)

evykassirer replied:
Oh actually there's a bug because I never made an "other users" version of this in the dark theme file! Exactly why variables are better :)

karlstolley commented on web/styles/right_sidebar.css:
```diff
@@ -13,11 +13,30 @@ $user_status_emoji_width: 24px;
+.toggle-narrow-users,
+.toggle-other-users {
+    width: 7px;
+    margin-top: 4px;
```
Looks like you can center this vertically by just setting `align-items: center` on the flex container and ditching the `margin-top` here. You might have to fiddle some with the height or line-height of the icon, but in testing just now, it looks pretty good to me.

evykassirer replied:
Thanks for noticing that, updated to do that instead!

karlstolley commented on web/styles/right_sidebar.css:
```diff
@@ -91,6 +110,28 @@ $user_status_emoji_width: 24px;
     }
 }
 
+.buddy-list-subsection-header {
```
I think this is where you'd set `align-items: center` to handle the arrow icon.

evykassirer commented on web/styles/right_sidebar.css:
```diff
@@ -13,11 +13,30 @@ $user_status_emoji_width: 24px;
+
+#narrow-users-presence-container,
+#other-users-presence-container {
+    &.no-display {
```
I'm a little nervous about global style rules for a classname that's that generic, but I'm open to ideas if you think something else would work better

evykassirer commented:
I've addressed and/or replied to all comments :)

evykassirer commented:
I need to do a followup to properly combine this work with #26196

evykassirer commented:
> I need to do a followup to properly combine this work with #26196

I've done this. Though there's a circular import being reported now and I'm not sure the best way to resolve it (activity_ui <> pm_list)

alya commented:
It's feeling good to me in manual testing! @timabbott do you want to take a look?

evykassirer commented on web/styles/right_sidebar.css:
```diff
@@ -13,20 +13,39 @@ $user_status_emoji_width: 24px;
-        padding-right: 15px;
-        padding-top: 1px;
-        padding-bottom: 2px;
+        /* Override .empty-list-message !important styling */
```
I don't like this, but I also don't have context on why we decided to use `!important` for the main empty list message.

I'm also assuming we don't want to use that styling here, but maybe we do?

evykassirer replied:
with the override:

<img width="292" alt="image" src="https://github.com/zulip/zulip/assets/5634097/a72cfe49-3129-4e66-8609-548b5d2ceecc">

without:

<img width="321" alt="image" src="https://github.com/zulip/zulip/assets/5634097/47a5df54-10e7-4032-9d99-82e5da92f4a1">

evykassirer commented:
repeating comment from above:

> There's a circular import being reported now and I'm not sure the best way to resolve it (activity_ui <> pm_list)

timabbott commented on web/src/buddy_data.js:
```diff
@@ -71,7 +84,26 @@ export function level(user_id) {
+    if (a_would_recieve_message && !b_would_recieve_message) {
+        return -1;
+    }
+    if (!a_would_recieve_message && b_would_recieve_message) {
```
"receive" is misspelled.

evykassirer replied:
thanks, fixed this!

timabbott commented:
This seems to introduce a lot of import cycles, which will need to be fixed before we can merge it. But for now it's just a linter error.

evykassirer commented:
> This seems to introduce a lot of import cycles, which will need to be fixed before we can merge it. But for now it's just a linter error.

See my notes above about this. I think it's all coming from one import cycle? Or at least the error was originally for just two files. I'm not sure what the best way to fix it is and would love some guidance.

timabbott commented:
Updating the chat.zulip.org deployment.

evykassirer commented:
most recent push fixes some failing unit tests

timabbott commented on web/src/pm_list.js:
```diff
@@ -162,13 +163,19 @@ export function handle_narrow_activated(filter) {
     update_private_messages();
+    if (activity_ui.user_filter) {
+        activity_ui.build_user_sidebar();
+    }
```
You main import cycle bug is that you shouldn't be adding this code here. `pm_list` is the module for the left sidebar "direct messages" section; it has nothing to do with the right sidebar. You can call your live-updates directly from `web/src/views_util.js` and `handle_post_view_change`.

evykassirer replied:
Oh of course, thanks

evykassirer replied:
Maybe I had thought this was a different `handle_message_view_deactivated` function...? Sorry for not realizing this

evykassirer replied:
Okay I think I got it a lot better, but now I have `views_util > activity_ui > inbox_ui > views_util`

timabbott commented on web/src/views_util.js:
```diff
@@ -43,6 +42,9 @@ export function show(opts) {
 
+    // This has to happen after resetting the current filter.
+    pm_list.handle_message_view_deactivated();
+
```
You presumably will want to revert moving this code when you fix the import cycle.

evykassirer replied:
done :)

timabbott commented on web/templates/right_sidebar.hbs:
```diff
@@ -18,7 +18,7 @@
             </div>
             <div id="buddy_list_wrapper" class="scrolling_list" data-simplebar>
-                <ul id="user_presences" class="filters" data-search-results-empty="{{t 'No matching users.' }}"></ul>
+                <ul id="narrow-user-presences" class="filters" data-search-results-empty="{{t 'No matching users.' }}"></ul>
```
What do you think about calling this `buddy-list-users-matching-view`? This aligns more closely with the string "no matching users", and elsewhere we're using "presence" mainly to refer to just the dots, not an entire user row.

evykassirer replied:
I thought "no matching users" is specifically about "matching" the search string, not the narrow -- so I would find this name more confusing.

If we want to remove presence, I'd be okay with `buddy-list-narrow-users`

timabbott replied:
Well that's why I put "matching view" in the name, not "matching filter".

I don't like the use of "narrow" because (1) the name is generally being phased out of the UI as a misnomer in favor of things like "messages view" or "message feed" and probably the code will follow and (2) we will likely be doing this in things like "recent conversations filtered to a stream", which does not match the historical definition of a "narrow".

evykassirer replied:
Ok sure :)

timabbott commented:
I extracted a couple prep commits as the series ending with https://github.com/zulip/zulip/pull/27402. I think https://github.com/zulip/zulip/pull/26717#discussion_r1373965942 would also be easy to merge as a prep commit after the rename proposed there.

evykassirer commented:
most recent push:

* rebase from main
* rename to `buddy-list-users-matching-view`
* fixed most of the import cycles, except that there's still one for `views_util > activity_ui > inbox_ui > views_util`

**remaining work**

Adding the "X inactive user" thing at the bottom of the lists

* I started a [conversation in #frontend](https://chat.zulip.org/#narrow/stream/6-frontend/topic/subscribed.20non-bot.20users/near/1671078) about ways to get those numbers without bots. 
* I'd love some design direction from @terpimost on this, and tagged him [in #design](https://chat.zulip.org/#narrow/stream/101-design/topic/buddy.20list.20split.20users/near/1669424) to ask for thoughts.
* Maybe we want the numbers to link to something?

The spacing in the user list is too cramped, so I'll look into that.

evykassirer commented on web/src/ui_init.js:
```diff
@@ -205,6 +205,36 @@ function initialize_right_sidebar() {
+            "fa-caret-left",
+            is_collapsed,
+        );
+    });
```
I was considering trying to set this up more generally instead of mostly copying the same code, but left it as this for now since I think it's more readable.

timabbott replied:
Yeah, this seems fine. It would be nice to move `initialize_right_sidebar` and `initialize_left_sidebar` into `sidebar_ui.js`. Let's plan to move those functions; could be in a prep PR if that's easy.

evykassirer replied:
#27484

evykassirer commented on web/src/buddy_list.js:
```diff
@@ -6,11 +6,14 @@ import render_presence_rows from "../templates/presence_rows.hbs";
 
 class BuddyListConf {
-    container_sel = "#buddy-list-users-matching-view";
+    narrow_container_sel = "#buddy-list-users-matching-view";
```
@timabbott would you prefer this be named as `matching_view_container_sel` or something?

timabbott replied:
`matching_view_container_selector` seems good. `sel` is an abbreviation and thus violates our coding standards.

evykassirer commented on web/src/buddy_list.js:
```diff
@@ -27,8 +30,10 @@ class BuddyListConf {
         const user_id = opts.key;
-        const $container = $(this.container_sel);
-        return $container.find(`${this.item_sel}[data-user-id='${CSS.escape(user_id)}']`);
+        const $narrow_users_container = $(this.narrow_container_sel);
```
Similar naming question here -- `$users_matching_view`?

timabbott replied:
`$users_matching_view_container` seems good

evykassirer commented on web/src/buddy_list.js:
```diff
@@ -80,12 +86,55 @@ export class BuddyList extends BuddyListConf {
         }
 
         const items = this.get_data_from_keys(more_keys);
+        const subscribed_users = [];
```
I also use "subscribed  users" language in a few places, maybe this should also be called `users_matching_view`?

timabbott replied:
I think `subscribed_users` seems correct when talking about the set of users themselves, and not the section in the UI where we display users that match the view.

evykassirer reviewed (commented):
Here are a few questions about naming :)

evykassirer commented:
Most recent pushes fix codecov issues

evykassirer commented:
most recent push:

* Includes an attempt to fix the circular import issue. I put it in its own commit because I'm not really sure if it makes sense to put the code where I did, but also I think it needs to be moved high up the call stack to not cause circular imports. Let me know if that looks okay!
* Adds a commit to show the "inactive user" message. There's still some design conversation happening about this on CZO

evykassirer commented:
This push is just fixing some test failures :)

evykassirer commented:
I feel like I'm doing a bit of whackamole 😅 fixed code coverage, now should have passing node tests and passing test coverage. Hoping this one passes CI 🤞

timabbott commented:
Opened https://github.com/zulip/zulip/pull/27457 for a couple prep commits.

evykassirer commented:
Updates:

* renamed a bunch of variables as discussed above
* I realized I wasn't updating the buddy list in a few places I needed to, so I added a bunch of function calls in the hashchange module. I don't really like this though, and would love ideas on other ways to update the sidebar

evykassirer commented:
I realized the inactive numbers were showing up during search, so I added a line of code to remove that.

evykassirer commented:
restating the remaining things:

* the "X inactive user" thing at the bottom of the lists needs design input, asked about that on CZO
* the last commit of this PR has some code I don't really like (but idk where else to put it to avoid circular imports) and I'm looking for input

evykassirer commented:
I updated the inactive user message commit to:
* Show the number of total users beside the section headings
* Include links to views to see subscribed users / all users, as discussed on [CZO](https://chat.zulip.org/#narrow/stream/101-design/topic/buddy.20list.20split.20users/near/1676057)

evykassirer commented:
Most recent updates, from conversation on CZO:

* remove the inactive user message
* drop the periods on the links to the user views
* indent the links
* hide the links when no users, still show it if we're showing all the relevant users for the list
* I renamed `matching_view_container_selector` to matching_view_list_selector` because there's a container around the list that's called "container" and that was confusing

alya commented:
Cool! I think we should count yourself in the DM conversation participants total. This looks odd:

<img width="232" alt="Screenshot 2023-11-06 at 10 47 10 AM" src="https://github.com/zulip/zulip/assets/2090066/555b1cfd-72c8-4ad5-b249-263d4f5e3bcf">

alya commented:
Perhaps this is known, but it seems like live-update does not work when a user is added to a stream, though it *does* work when they are removed.

alya commented:
This is an edge case, but when a stream has no users, I think it should say "None." (just like the "Others" section). Currently, I see this instead:

<img width="307" alt="Screenshot 2023-11-06 at 10 53 40 AM" src="https://github.com/zulip/zulip/assets/2090066/e45ab468-ca89-47de-a678-e524d8121cb5">

alya commented:
I don't know what the conditions for this are, but I'm seeing the "View all subscribers" link pointing to the wrong stream (the URL in the bottom left is from hovering over that link):

<img width="1415" alt="Screenshot 2023-11-06 at 10 56 30 AM" src="https://github.com/zulip/zulip/assets/2090066/c80fb318-5f1e-4da7-a442-04b436fdfe7a">

evykassirer commented on web/src/sidebar_ui.js:
```diff
@@ -181,4 +182,43 @@ export function initialize_right_sidebar() {
+    // There's no way to get to the subscriber view by url,
+    // so we have this click handler to change the tab.
+    // Ideally in the future we have a url structure to go
+    // directly to the subscriber settings without this.
```
Discussion about that on CZO [here](https://chat.zulip.org/#narrow/stream/6-frontend/topic/open.20stream.20subscribers/near/1676211)

evykassirer commented:
Thanks for catching these Alya!

* Added self to the count of PM participants
* Fixed the link pointing to the wrong stream
* Made the empty list situation work the same for subscribed users as other users

<img width="222" alt="image" src="https://github.com/zulip/zulip/assets/5634097/3c856616-f500-40c0-a572-007df0c967e7">


> It seems like live-update does not work when a user is added to a stream, though it does work when they are removed.

I noticed it didn't work when a user is added to a stream, sorry for not pointing it out explicitly and asking about it. I've been trying to keep the PR smaller since it's already quite unwieldy, and was thinking this could be a follow up task, but especially with small lists of subscribers I can see an argument for including live-update right away. Do you have a preference here?

alya commented:
Thanks!

> I noticed it didn't work when a user is added to a stream, sorry for not pointing it out explicitly and asking about it. I've been trying to keep the PR smaller since it's already quite unwieldy, and was thinking this could be a follow up task, but especially with small lists of subscribers I can see an argument for including live-update right away. Do you have a preference here?

I'll let @timabbott answer this one.

karlstolley commented on web/styles/right_sidebar.css:
```diff
@@ -13,11 +13,37 @@ $user_status_emoji_width: 24px;
+
+#buddy-list-users-matching-view,
+#buddy-list-other-users {
+    max-width: 90%;
```
Should this just be a `width` value? `max-width` set to a percentage is usually the road to frustration, and doesn't differ appreciably from plain old `width`

evykassirer replied:
It's been too long since I wrote this to remember 😅 I'll take your word for it and switch to `width` though, looks okay for now

karlstolley commented on web/styles/right_sidebar.css:
```diff
@@ -86,19 +112,60 @@ $user_status_emoji_width: 24px;
     .empty-list-message {
-        font-size: 1.2em;
+        font-style: italic;
+        color: hsl(0deg 0% 67%);
```
It might be nice to put this in a CSS variable, too, along with the other colors.

evykassirer replied:
Yeah sure! I was grabbing it from the old empty message styling

karlstolley commented:
I did a pass of the CSS here. Looks good minus a couple of things I pointed out.

As a side note, I spent some time just walking through the DOM, inspecting the right sidebar and its styles. There's lots that can be tightened and cleaned up there, but all outside of this PR.

evykassirer commented:
This push has these changes:

* I squashed some of the temp commits together, since I wasn't sure if that was helping with review and wanted to simplify.
* The number of users in the headings shortens when there are >1k users, and the exact number of users is always visible in a tooltip
* Clicking the toggle no longer closes the compose box
* "No active users" is shown instead of "None" when there are inactive users but none visible
* Did some refactoring to share code between various header/empty-message/etc functions
* Fixed circular imports in a different way, calling the update functions from a better places, fixing issues where the sidebar wasn't always updating during narrowing actions (still separating it as the last commit in this PR for easier review)

alya commented:
Let's show a "View all users" link when there are no sections (e.g., Inbox), and not all users are visible in the sidebar.

<img width="285" alt="Screenshot 2023-11-15 at 10 33 56 PM" src="https://github.com/zulip/zulip/assets/2090066/4800ef57-751b-46dc-8dfa-718f8046db11">

alya commented:
I think in that situation, we can also change the tooltip on the USERS heaading to give the exact number of users. The magnifying glass will still have the "Search people" tooltip.

<img width="170" alt="Screenshot 2023-11-15 at 10 34 53 PM" src="https://github.com/zulip/zulip/assets/2090066/49ebefd4-0b0e-4dd5-a935-9a87947ef366">

alya commented:
Let's change the tooltip on the "Others" section to "N other users". After playing with it, I think it'll be more clear.

alya commented:
We also need to handle plurals properly (including internationalization -- mentioning that just in case):

<img width="100" alt="Screenshot 2023-11-15 at 10 39 09 PM" src="https://github.com/zulip/zulip/assets/2090066/3b5b3eff-cd43-4bdf-b5a8-da429380a512">

andersk commented on web/src/peer_data.ts:
```diff
@@ -69,9 +69,15 @@ export function potential_subscribers(stream_id: number): User[] {
-    const subscribers = get_user_set(stream_id);
-    return subscribers.size;
+export function get_subscriber_count(stream_id: number, include_bots: boolean = true): number {
+    let subscribers = [...get_user_set(stream_id).keys()];
```
This has negated the whole point of the `LazySet` optimization by copying the entire set to an array every time.

andersk commented on web/src/peer_data.ts:
```diff
@@ -69,9 +69,15 @@ export function potential_subscribers(stream_id: number): User[] {
+
+    if (!include_bots) {
+        const active_user_ids = people.get_realm_active_human_user_ids();
+        subscribers = subscribers.filter((user_id) => active_user_ids.includes(user_id));
```
This takes quadratic time.

evykassirer replied:
Fixed this after some conversation on [CZO](https://chat.zulip.org/#narrow/stream/101-design/topic/buddy.20list.20split.20users/near/1683522)

evykassirer commented:
Updates:

* Removed quadratic thing that Anders pointed out
* Show a "View all users" link when there are no sections (e.g., Inbox), and not all users are visible in the sidebar.
* Changed the tooltip on the "Others" section to "N other users"
* Handle plurals properly

Remaining:

* Change the tooltip on the USERS heading to give the exact number of users (probably keeping that there all the time)
* I realize I still haven't addressed the issue of updating the buddy list when someone is subscribed or unsubscribed. I'm wondering if we can keep this as a followup though.

(and remaining things from before)

* Merging #27637, which is a prep PR for this one (and then integrating that into this PR)
* Anders had some feedback about the triangle on [CZO](https://chat.zulip.org/#narrow/stream/101-design/topic/buddy.20list.20split.20users/near/1682960) that's still under discussion.
* The tests in this PR are currently failing. I think I've been spending too much time on cleaning up tests as I go, with the product back-and-forth that's going on, so I want to focus on getting the product pieces addressed before worrying about fixing unit tests or code coverage for now.

evykassirer commented:
Updates:

* Tooltip on the USERS heading now gives the exact number of users (it always shows this, and the search icon has been left unchanged)
* The buddy list is rerendered when someone is subscribed or unsubscribed.

Remaining:

* Merging #27637, which is a prep PR for this one
* Figuring out what we want with the triangle
* Fixing failing tests once everything else feels finalized (or close enough)

evykassirer commented on web/src/activity_ui.js:
```diff
@@ -22,6 +21,9 @@ import * as util from "./util";
 export let user_filter;
 
+// Function initialized from `ui_init` to avoid importing narrow.js and causing circular imports.
+let narrow_by_email;
```
@timabbott can you confirm if this commit works in the way you'd like? I'll integrate it into the main commit once you've given feedback.

timabbott replied:
Yep, this is fine.

evykassirer commented:
Pushing to resolve conflicts. @alya let me know if there's anything left on the product-side for this before I fix up the unit tests!

Locally one thing I'm noticing is a lag in showing the "No active users" message, and I'll look into it in a bit.

alya commented:
How does the rounding function work? We might want to play with it (e.g., I think I'd prefer 1.5K to 1K for 1509 users), but that should definitely be a separate follow-up.

<img width="189" alt="Screenshot 2023-11-19 at 10 47 03 PM" src="https://github.com/zulip/zulip/assets/2090066/33bdb31f-0e68-4079-9fb4-c2ac544c6084">

alya commented:
This tooltip (in a large window) looks cut off:

<img width="204" alt="Screenshot 2023-11-19 at 10 55 33 PM" src="https://github.com/zulip/zulip/assets/2090066/665c3037-6a21-4a61-adcd-39abe3a0da8c">

alya commented:
OK, this one might be controversial, so let's see how the discussion [here](https://chat.zulip.org/#narrow/stream/101-design/topic/stream.20subscribers.20in.20top.20bar/near/1685517) shakes out.

> Could we add the name of the stream to the tooltip as well, for clarity? I think "17 subscribers in # stream" (with the correct privacy icon) should work. (Note that we just did something similar in https://github.com/zulip/zulip/pull/27666 .)

timabbott commented on web/src/buddy_list.js:
```diff
@@ -10,10 +10,10 @@ import * as padded_widget from "./padded_widget";
+    container_selector = "#buddy-list-users-matching-view";
+    scroll_container_selector = "#buddy_list_wrapper";
+    item_selector = "li.user_sidebar_entry";
+    padding_selector = "#buddy_list_wrapper_padding";
```
Extracted this first commit, which lgtm, as https://github.com/zulip/zulip/pull/27807.

evykassirer commented on web/src/buddy_data.js:
```diff
@@ -32,10 +32,18 @@ let is_searching_users = false;
-    const empty_list_message = is_searching_users
-        ? $t({defaultMessage: "No matching users."})
-        : $t({defaultMessage: "None."});
+export function update_empty_list_placeholders(
```
Darn, this commit is actually not quite right because the call to `update_empty_list_placeholders` in this file doesn't have the arguments to pass to it. So this will break if "No active users" is shown and the user searches and then stops searching.

I can't think of a cleaner way to do this than to have variables in this file that keep track of if there are inactive users, or I guess a function that calculates those variables from scratch (but from a separate module that's imported from both here and buddy list, to avoid circular imports)? Thoughts?

evykassirer replied:
Aha I figured out a solution, which makes everything nicer. I realized that the `populate` function is called every time the search string changes, to repopulate how the users are displayed, so I can just do the empty list stuff only from the populate function. I'll make some changes and push that.

evykassirer commented on web/src/buddy_list.js:
```diff
@@ -1,16 +1,23 @@
 import $ from "jquery";
 
+import render_section_header from "../templates/buddy_list/section_header.hbs";
```
This commit ("buddy list: Add user count and link to see full user list.") would ideally be separated into three commits:

* The refactor to get the data object
* Making the template for the headers and putting the user count in it
* Adding the links

I can do this in a bit.

evykassirer replied:
Done!

evykassirer commented on web/src/buddy_list.js:
```diff
@@ -79,51 +86,106 @@ export class BuddyList extends BuddyListConf {
+                user_count: message_view_header.get_formatted_sub_count(other_users_count),
+                tooltip_text: $t(
+                    {defaultMessage: "{N, plural, one {# other user} other {# other users}}"},
+                    {N: other_users_count},
```
Maybe it would be better to move the `plural` thing into handlebars than here?

timabbott replied:
I expect so.

evykassirer replied:
Oh actually maybe not, because we need to differentiate between "participant", "user", and "subscriber". I'll leave it as is for now

evykassirer commented on web/src/message_view_header.js:
```diff
@@ -10,6 +10,14 @@ import * as recent_view_util from "./recent_view_util";
 import * as rendered_markdown from "./rendered_markdown";
 import * as search from "./search";
 
+export function get_formatted_sub_count(sub_count) {
```
It seems like this function is longer being used in this file, so I could move it to `buddy_list` -- though I want to look into why it's no longer being used for the navbar tooltip, since the idea was to have similar behavior.

timabbott replied:
I think the tooltip now just shows the actual number, so that function was removed as unused. 

If you need it for the buddy list, you should add a new copy where you'll use it.

evykassirer replied:
Moved it

evykassirer reviewed (commented):
I went through commit by commit, since it's been a while since I took a look at the wide bigger picture, and left several comments. There are some things I think I should definitely change and others that are more questions. @timabbott not sure if it makes sense for you to wait on some of that before doing a review or not.

timabbott commented on web/src/activity_ui.js:
```diff
@@ -158,7 +162,7 @@ export function narrow_for_user_id(opts) {
     const email = person.email;
 
-    narrow.by("dm", email, {trigger: "sidebar"});
+    narrow_by_email(email);
```
We should expect it to always be set. If we wrote something, it'd be an `assert` -- one should prefer asserts rather than writing `if` branch for a case that is not possible. You can use `import assert from "minimalistic-assert";` -- we mostly do it in TS files since it does type narrowing for us, but we use it in JS in at least one place.

timabbott replied:
This `narrow_by_email` extraction would be a great prep commit/PR BTW -- it's distracting when reading the main commit.

evykassirer replied:
Great idea, I'll make it a prep commit right now :)

evykassirer replied:
Noting here that I chose not to add the assert because it was making writing tests more annoying (I had to make sure it was initialized) and I'm running low on time for working today so I'm just gonna go without the assert for now.

timabbott commented on web/src/buddy_data.js:
```diff
@@ -250,7 +306,15 @@ function maybe_shrink_list(user_ids, user_filter_text) {
+    // We want to always show PM recipients even if they're inactive.
+    const pm_ids_string = narrow_state.pm_ids_string();
+    const pm_ids_list = pm_ids_string ? people.user_ids_string_to_ids_array(pm_ids_string) : [];
+    const pm_ids_set = new Set(pm_ids_list);
```
Yes, definitely.

evykassirer replied:
Great, done

timabbott commented on web/src/buddy_data.js:
```diff
@@ -24,8 +30,35 @@ import * as util from "./util";
+        : $t({defaultMessage: "None."});
+
+    $("#buddy-list-other-users").data("search-results-empty", empty_list_message);
+    if ($("#buddy-list-other-users .empty-list-message").length) {
```
OK. I don't love using `.data("search-results-empty", empty_list_message)` as a setter; prefer `.attr("data-search-results-empty", ...)`. The `.data` function is evil in that it reads a `data-search-results-empty` field, but writes a thing that is only read by `jQuery`, so any use of this is both harder to grep (as you might have a `data-search-results-empty` in the DOM and also makes migrating off `jQuery` harder.

Posted about https://chat.zulip.org/#narrow/stream/6-frontend/topic/jQuery.20.2Edata/near/1686087 since we should just discuss and potentially lint-ban this as a project.

evykassirer replied:
When I change it, it doesn't work anymore. I'll follow up on CZO.

timabbott commented on web/styles/right_sidebar.css:
```diff
@@ -93,7 +110,13 @@ $user_status_emoji_width: 24px;
+        /* Overwrite default empty list font size, to look better under the subheaders. */
+        font-size: 14px;
+        /* Override .empty-list-message !important styling */
+        padding: 0 !important;
```
I don't understand this comment -- are you saying this `!important` overrides another `!important` somewhere else?

timabbott replied:
Ahh, yeah. I suspect the base `empty-list-message` base CSS class doesn't super make sense; can you start a #frontend discussion about how we should refactor that?

evykassirer replied:
https://chat.zulip.org/#narrow/stream/6-frontend/topic/empty-list-message.20styling

evykassirer commented on web/src/buddy_list.js:
```diff
@@ -216,22 +359,28 @@ export class BuddyList extends BuddyListConf {
+        const new_pos_in_all_users = opts.new_pos_in_all_users;
         const html = opts.html;
-        const pos = opts.pos;
+        const key_following_insertion = opts.key_following_insertion;
```
(did that)

timabbott commented on web/src/buddy_list.js:
```diff
@@ -54,15 +67,20 @@ class BuddyListConf {
 
 export class BuddyList extends BuddyListConf {
-    keys = [];
+    all_user_ids = [];
```
Yes that sort of split commit would be helpful.

evykassirer replied:
Done :)

evykassirer commented:
I've made a bunch of technical updates to make it easier for Tim to review. I've also opened a prep PR.

Remaining and unaddressed questions:

* Addressing Alya's most recent design/product feedback
* [starting this conversation in #frontend](https://github.com/zulip/zulip/pull/26717#discussion_r1399937533)
* Unit tests

evykassirer commented:
I'll maintain the TODO list in the description of the PR from now on.

timabbott commented:
Redeploying this on chat.zulip.org.

evykassirer commented on web/src/buddy_list.js:
```diff
@@ -86,6 +88,77 @@ export class BuddyList extends BuddyListConf {
     all_user_ids = [];
     users_matching_view_ids = [];
     other_user_ids = [];
+    tooltips_initialized = false;
```
I don't really like this, but I couldn't figure out a different way to initialize the tooltip only once while still taking advantage of shared code through functions like `this.total_subscriber_count`. Would it better to just totally copy that logic and put this in e.g. `click_handlers` with the other tippy code?

timabbott replied:
The usual way to do this would be to have a `buddy_list.initialize` function called by `ui_init.js` -- that's how all our other modules are initialized. You could put that just before or just after `activity_ui.initialize` to control when it happens; seems like before would be more correct?

Oh or another totally reasonably plan would be for `initialize_right_sidebar` in `sidebar_ui.js` to call `initialize_tooltips`.

evykassirer replied:
Ah of course, thanks!

evykassirer commented on web/src/buddy_list.js:
```diff
@@ -86,6 +88,77 @@ export class BuddyList extends BuddyListConf {
+                // This will default to "bottom" placement for this tooltip.
+                placement = "auto";
+            }
+            const subscriber_count_function = this.total_subscriber_count;
```
Because `this` is different inside `onShow`. I've seen `that = this` things, but find that confusing. Is this a reasonable way of doing this?

karlstolley replied:
I think that's okay. It might read better to set the variable to `get_total_subscriber_count`, though.

evykassirer replied:
cool, did that

evykassirer commented on web/src/buddy_list.js:
```diff
@@ -86,6 +88,77 @@ export class BuddyList extends BuddyListConf {
+                            {N: other_users_count},
+                        );
+                    }
+                    instance.setContent(tooltip_text);
```
I opted for generating the text here instead of in the template because the wording is different depending on which of the three cases we're in. But it looks fairly ugly with the nested ifs -- is there a nicer way to do this?

karlstolley replied:
It seems like you ought to be able to make the assignment to `tooltip_text` just the once. The logic in any if/else statements seem like the should be limited to determining the appropriate `N` and the singular/plural noun forms. Maybe just constructing an object literal that gets interpolated into `defaultMessage` and `N`'s value? `{ N: <val>, singular: <val>, plural: <val> }? I don't know if that'd help, but maybe that gives you some alternate technique to stumble onto another one.

evykassirer replied:
Hm yeah I'm not sure if that would simplify it significantly, and I'm also not sure if that would make the translation strings harder to translate -- I think the words need to be shown in the context phrase that they're seen in. I think leaving it like this is best for now.

timabbott replied:
I think this is probably fine as written, resolving. I'm pretty sure the translation library needs a direct `$t` call without indirection to find strings.

evykassirer commented on web/templates/buddy_list/section_header.hbs:
```diff
@@ -1,4 +1,4 @@
-<h5 id="{{id}}" class="buddy-list-heading no-style tippy-zulip-tooltip hidden-for-spectators" data-tippy-content="{{tooltip_text}}" data-tippy-placement="bottom">
+<h5 id="{{id}}" class="buddy-list-heading no-style hidden-for-spectators">
```
I took the tooltip out of here and moved to the logic above because 

(1) left-alignment wasn't working because the tooltip was attaching to a nearby html element instead of `body` and didn't have space 

(2) I wanted to add the delay

But maybe there's a simpler way of making this change?

karlstolley replied:
Hmmmm. I'm not sure I'm following your comment here based on the code in this area. It might be worth studying how delays and tooltips are set on different headings in the left side bar, but I don't trust that that's the help you're looking for.

evykassirer replied:
Essentially there's a way of doing tooltips with `tippy-zulip-tooltip` and `data-tippy-content`, but it didn't seem to support the two things I needed above, so I switched to the code in `buddy_list.initialize_tooltips`. I think the left sidebar code isn't significantly different from what I ended up with here, so I'll leave it as is for now and let @timabbott share if he wants anything different from this commit.

evykassirer reviewed (commented):
I tried making the tippy changes but I'm not really confident I did it the right way -- I'm leaving some comments on things I wasn't sure about, for review.

evykassirer commented:
Just updated the unit tests and coverage! I removed `buddy_list.js` coverage in a commit in the middle of this PR, and did a lot of tweaking to update unit tests. Hopefully this PR is close to ready now :)

karlstolley commented on web/src/buddy_list.js:
```diff
@@ -55,16 +86,208 @@ class BuddyListConf {
+                // Because the buddy list subheadings are potential click targets for purposes
+                // having nothing to do with the subscriber count (collapsing/expanding), we
+                // delay showing the tooltip until the user lingers a bit longer.
+                delay: [300, 0],
```
Rather than hardcode this value, it would be better to use something like the `INTERACTIVE_HOVER_DELAY` constant from `tippyjs.ts`. I think there's a goal to have a relatively small set of delays that are used uniformly throughout the UI.

evykassirer replied:
Sure, I'll try that! For the record @alya this means that the tooltip would show up after 425ms of hover and go away 20ms after hover ends.

evykassirer commented:
I addressed Karl's comments and rebased to resolve merge conflicts. 

@timabbott, Karl said on CZO that "once the merge conflicts have been resolved, the next step is more CZO burn-in"

timabbott commented:
Test-deploying on chat.zulip.org for further testing.

timabbott commented on web/src/buddy_list.js:
```diff
@@ -278,6 +278,7 @@ export class BuddyList extends BuddyListConf {
                 header_text,
                 user_count: get_formatted_sub_count(subscriber_count),
                 toggle_class: "toggle-users-matching-view",
+                is_collapsed: $("#buddy-list-users-matching-view").hasClass("collapsed"),
```
Is there a reason to read this from the DOM rather than making a global variable to keep track of that state? I guess it doesn't really matter, but it feels like this module should be keeping track of whether that is collapsed for itself, rather than using the DOM as its data storage.

evykassirer replied:
My general preference is to avoid having multiple sources of truth, i.e. a variable could end up out of sync with the DOM. The main reason I'd keep a variable is if there's concerns about performance or something, which I don't have here. I'm open to hearing more reasons to change it though!

timabbott replied:
I see the global variable as a dedicated place to store "Should this be collapsed" and the DOM state as "is this currently collapsed?" 

The approach you have here is very fragile, in that you're reading this state after code earlier has emptied the adjacent containers, and it's not obvious to me that this code as written even works -- didn't we just delete from the DOM the thing we're reading from at the top of this function?

Maybe it's in a slightly different bucket, but even if this turns out to work as written, it's fragile and hard to verify. What is not fragile and hard to verify is having a single source of truth, a variable that tracks the state, and rendering logic that aligns with it.

evykassirer replied:
Okay, I see. Thanks for explaining.

timabbott commented:
Posted one comment on the new 3 fixup commits; otherwise they lgtm and should be fine to squash in.

evykassirer commented on web/src/buddy_list.js:
```diff
@@ -55,16 +87,210 @@ class BuddyListConf {
+        }
+    }
+
+    update_empty_list_placeholders({has_inactive_users_matching_narrow, has_inactive_other_users}) {
```
`has_inactive_users_matching_narrow` should be `has_inactive_users_matching_view`

evykassirer commented:
I'll squash the commits in once we've resolved that last comment thread

evykassirer commented:
* squashed one of the recent three commits into the main commit
* squashed the other two together, still temporarily at the end of the PR, now with variables to store collapsed state
* current investigating [this bug](https://chat.zulip.org/#narrow/stream/464-kandra-js-errors/topic/Error.3A.20Unknown.20user_id.20in.20get_by_user_id.3A.2029815) that's probably related to this PR

evykassirer commented:
Error is addressed now in the last commit. Ready for review before a final squash and merge!

timabbott commented on web/src/activity_ui.js:
```diff
@@ -93,13 +93,14 @@ export function build_user_sidebar() {
     }
 
-    const filter_text = get_filter_text();
+    const filter_text = user_filter ? user_filter.text() : "";
```
Why is this change correct -- is it to avoid the `get_filter_text` blueslip warning? I'm wondering if there's a deeper issue with `activity_ui.initialize` being called too late that you're masking with this change.

evykassirer replied:
This is the trace where this comes up:

```js 
    build_user_sidebar activity_ui.js:97
    handle_post_view_change narrow.js:1034
    activate narrow.js:533
    do_hashchange_normal hashchange.js:159
    hashchanged hashchange.js:411
    initialize hashchange.js:424
    initialize_everything ui_init.js:669
    js ui_init.js:740
```

So it's from hashchange's initialize function. I'm going to add another commit after this main one to move the `hashchange` initialization after `activity_ui`, which doesn't cause tests to fail and hopefully won't cause other issues.

timabbott commented on web/src/buddy_list.js:
```diff
@@ -1,16 +1,38 @@
+        sub_count = Number.parseInt(sub_count / 1000, 10) + "k";
+    }
+    return sub_count;
+}
```
This should possibly be using this module:

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/NumberFormat/NumberFormat

I've not checked `caniuse` etc., or played with it, but the `engineering` style might be the one that spells values this way?

evykassirer replied:
Oh yeah this is bad form that's not new here, it's moved from elsewhere in the file. I fixed it when going over typescript. I could just use `Math.floor`, which is what I did there, which feels simplest here and keeps the same logic as was already there.

timabbott replied:
This function is new in this PR, I'm pretty sure. But the main point is internationalization -- I'm not confident every language spells numbers that way.

evykassirer replied:
Ohhh I think maybe I was copying this from other code that did something similar then, that's since been deleted. I'm just pretty sure I didn't actually write that code. Sorry for not noticing its jankness when I copied it.

That makes sense about internationalization. I'll look into this!

evykassirer replied:
I asked a question about this [here](https://chat.zulip.org/#narrow/stream/101-design/topic/buddy.20list.20split.20users.20.2326717/near/1734019)

evykassirer replied:
Updated to use `new Intl.NumberFormat(user_settings.default_language, {notation:  "compact"}).format(sub_count);`

timabbott commented on web/src/buddy_list.js:
```diff
@@ -27,8 +49,18 @@ class BuddyListConf {
+        const $other_users_container = $(this.other_user_list_selector);
+        return $other_users_container.find(
+            `${this.item_selector}[data-user-id='${CSS.escape(user_id)}']`,
+        );
```
This code seems unnecessarily duplicative; couldn't this just use a container containing both elements with a single `$container.find(\`${this.item_selector}[data-user-id='${CSS.escape(user_id)}']\`)`?

evykassirer replied:
Oh true! Thanks :)

timabbott commented on web/src/buddy_list.js:
```diff
@@ -55,16 +87,259 @@ class BuddyListConf {
+                // Don't show tooltip on touch devices (99% mobile) since touch pressing on users in the left or right
+                // sidebar leads to narrow being changed and the sidebar is hidden. So, there is no user displayed
+                // to show tooltip for. It is safe to show the tooltip on long press but it not worth
+                // the inconvenience of having a tooltip hanging around on a small mobile screen if anything going wrong.
```
These comment lines should be wrapped to more like 80 characters. We allow longer lines for URLs and  places where it's more readable that way, but this is definitely not more readable written this way.

evykassirer replied:
Fixed!

timabbott commented on web/src/stream_events.js:
```diff
@@ -209,6 +213,6 @@ export function process_subscriber_update(user_ids, stream_ids) {
     }
-    compose_fade.update_faded_users();
     user_profile.update_user_profile_streams_list_for_users(user_ids);
+    activity_ui.build_user_sidebar();
```
Can this logic be conditional on a stream being modified being the one you're looking at? This rebuild is a somewhat expensive operation, and in a big organization, the `peer_add` events (which are decently common) will usually be related to a stream that isn't your current view (because DMs, inbox, recent, all messages never match, and there's probably a lot of streams, most of which are irrelevant).

timabbott replied:
Also please move this commit to the end -- and I think you can squash all the other older fixup commits except this and the `compose_fade` one (which is a nice atomic feature removal), following rebasing on top of https://github.com/zulip/zulip/pull/28848.

evykassirer replied:
To clarify -- you want the PR to be [every commit except this commit and the compose_fade one, squashed together] and then [the compose_fade commit] and then [this commit] so three total? Or do you want more separation than that still?

evykassirer replied:
> Can this logic be conditional on a stream being modified being the one you're looking at? 

Yup, updated the commit! Also moved it to the end

timabbott commented on web/tests/peer_data.test.js:
```diff
@@ -235,6 +242,11 @@ test("get_subscriber_count", () => {
+    peer_data.add_subscriber(india.stream_id, welcome_bot.user_id);
+    assert.deepStrictEqual(peer_data.get_subscriber_count(india.stream_id), 2);
+    // Get the count without bots
+    assert.deepStrictEqual(peer_data.get_subscriber_count(india.stream_id, false), 1);
```
These tests were in the wrong commit, I fixed that in #28848 which should let us merge the changes to several modules not related to the buddy list that this PR requires.

timabbott commented on web/src/buddy_list.js:
```diff
@@ -55,16 +87,259 @@ class BuddyListConf {
+        this.other_user_ids = [];
+
+        // Reset data to be relevant for this current view.
+        this.set_render_data();
```
So `this.set_render_data` seems to inspect values like `users_matching_view_ids`, both directly and indirectly (via `should_hide_headers`); what's the thinking of calling this with stale/bogus data at this point?

evykassirer replied:
Oh shoot yeah, it should be calculating `should_hide_headers` here. I think everything else doesn't use the bogus data. The way to guarantee this would be to pull the data calculation fully out of the BuddyList class, along with `total_subscriber_count`. I'll do that now.

evykassirer replied:
Pushing the last commit with that change. I'll come back to look at your other code review comments later (maybe tomorrow), gotta head out for a bit now

timabbott replied:
That refactor looks like an improvement.

timabbott commented on web/src/buddy_list.js:
```diff
@@ -55,16 +87,259 @@ class BuddyListConf {
+            $("#buddy-list-users-matching-view").empty();
+            $("#buddy-list-users-matching-view").append(empty_list_widget);
+        }
+    }
```
This function is very hard to read. I think you should just have a single block of conditionals at the top that sets two variables, `matching_view_empty_list_message` and `other_users_empty_list_message` and then does something with those. As it is, I think you're doing some sort of interesting fallthrough approach between the two blocks of conditionals that is just really hard to reason about -- and every time I read this I expect the `empty_list_message` calculated at the start to be used for the matching/active section, but then you're doing it in the other order.

evykassirer replied:
Ah yeah I see what you mean. I've updated the `buddy list: Show 'No active users' for empty lists with inactive users.` commit to calculate the two empty list messages first, and then start with setting matching users (instead of other users).

evykassirer replied:
If we want to keep this commit separate, I probably should also swap some ordering in an earlier commit, but I also think it would probably make sense to just squash this into the main buddy list commit.

timabbott replied:
Yeah just squash it.

timabbott commented on web/src/buddy_list.js:
```diff
@@ -97,45 +408,161 @@ export class BuddyList extends BuddyListConf {
+        // then we don't show section headers and only show one untitled
+        // section.
+        return this.users_matching_view_ids.length === 0 && !current_sub && !pm_ids_set.size;
+    }
```
Hmm, why is this the right check? I guess it would seem to me like the check should be just `!current_sub && !pm_ids_set.size`, without checking `this.users_matching_view_ids.length === 0`, which is by definition going to be true if the other things are true?

evykassirer replied:
Ah true! I think this happened because I started with just `this.users_matching_view_ids.length === 0` and adjusted it later. Thanks for noticing that. I'll edit the `Calculate render data once per view.` commit to calculate `should_hide_headers` that way.

timabbott commented on web/src/buddy_list.js:
```diff
@@ -97,45 +422,148 @@ export class BuddyList extends BuddyListConf {
+        const $other_user_elems = $other_user_obj.map((_i, elem) => $(elem));
+
+        return [...$users_matching_view_elems, ...$other_user_elems];
     }
```
I guess this isn't new, but do you understand the purpose of the `.map` here? I would have thought that we already had jQuery objects in the `.find` result.

Hmm is this function dead code? I don't see any relevant matches to `git grep get_items`.

evykassirer replied:
Great catch, it looks like it became dead in the `compose fade users: Remove fade logic in favor of buddy list split` commit. I'll another commit directly after it to remove `get_items`.

evykassirer commented:
Addressed some comments locally. Will address more later, and push once everything's addressed.

evykassirer commented:
Pushing what I have so far. Remaining issues I know of:

[ ] `get_formatted_sub_count` number formatting -- I asked some followups on CZO
[ ] profiling before and after `buddy list: Calculate render data once per view.` -- I need to look a bit into how to do this properly, but plan to do that soon
[ ] squashing commits -- I'm biasing to not squashing right away because it maybe helps you see what I've changed in such a huge PR, and it's harder to unsquash than to squash, but once you clarify what you want squashed I can do that

evykassirer commented:
Just the profiling piece left, but feeling a bit too tired today for learning a new thing. I'll probably get to it some time this weekend.

timabbott commented:
Looks like you've got a puppeteer failure as well.

evykassirer commented:
I tried using some profiling tools but wasn't able to see a notable change. I might be using them wrong though, so I'll try to get someone to pair with me on it later this week

timabbott commented:
OK rebased this, did the squash, and another review pass -- looks great, marked to merge once CI passes.

evykassirer commented:
Ah oops. We could also revert the sidebar typescript change, merge this, merge #29022, and then convert sidebar after that.

I'm pretty open to any path going forward -- do you have a preference?

timabbott commented:
Reverting is probably more annoying than this approach, I think:

> Another option would be to just put the code you're adding to sidebar_ui.js into a buddy_list.initialize method.

So I don't think we should consider that option. Probably we can just merge via https://github.com/zulip/zulip/pull/29022.

timabbott commented:
Merged via https://github.com/zulip/zulip/pull/29022.
